packages/extension-browser/src/devtools/panel/utils/network.ts

```diff
@@ -0,0 +1,148 @@
+import { Entry, Log } from 'har-format';
+
+import { browser } from '../../../shared/globals';
+import { mapHeaders } from '../../../shared/headers';
+
+import { sendMessage } from './messaging';
+
+/** Track the number of redirects by request. */
+const hops = new Map<string, string[]>();
+
+/**
+ * Generate `fetch::start` events from `devtools.network.onRequestFinished`.
+ * These are forwarded to the content-script via the background-script.
+ *
+ * TODO: Find an event which can be used to trigger these earlier.
+ */
+const generateFetchStart = (request: chrome.devtools.network.Request) => {
+    sendMessage({ fetchStart: { resource: request.request.url } });
+};
+
+/**
+ * Wait for the provided entry to include all response data.
+ * This works around a bug in Firefox which sometimes triggers
+ * onRequestFinished before HAR entries have been populated.
+ * https://bugzilla.mozilla.org/show_bug.cgi?id=1472653
+ */
+const waitForFullEntry = (entry: Entry, retries = 5): Promise<Entry> => {
+    return new Promise((resolve) => {
+
+        // Don't wait if the entry has already been populated.
+        if (typeof entry.response.content.size === 'number' && typeof entry.response.headersSize === 'number' && entry.response.status) {
+            resolve(entry);
+
+            return;
+        }
+
+        // Give up after a few attempts to avoid retrying forever.
+        if (!retries) {
+            browser.devtools.inspectedWindow.eval(`console.warn("[webhint] HAR missing data after max retries for ${entry.request.url}")`);
+            resolve(entry);
+
+            return;
+        }
+
+        // Wait just a bit to give the HAR time to be populated.
+        setTimeout(() => {
```

Comment:
Await `hint/src/lib/utils/misc/delay` here (will avoid the need for `new Promise` above).

Reply:
Actually still needed the new Promise because of the callback from `getHAR`. Going to skip using `delay` for now.

Reply:
ok!